File: Leetcode75/98. Validate Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public boolean check(TreeNode node, TreeNode upper, TreeNode lower){
        if(node == null) return true;
        
        /*Remove special cases that the node's value = MAX or MIN at the beginning tree
            lower: null max  upper: null      OR:            max                                                                                                                         /                                         
                    /         \                            max
                   2           3
        when the node.val = Min or Max, it is only true if the node is the root of the tree,
        otherwise, it will be false. And when the node is the root, upper = lower = null.
        And other cases, if the node is not a root but has the value = MIN or MAX, it has
        to be false.
        */
        if(upper != null && node.val >= upper.val) return false;
        if(lower != null && node.val <= lower.val ) return false;
        
        boolean left, right;
        left = check(node.left, node, lower);
        right = check(node.right, upper, node);
        return left&&right;
    }
    
    public boolean isValidBST(TreeNode node) {
        if(node == null) return true;
        TreeNode upper = null, lower = null;
        return check(node, upper, lower);
    }   
}

/**/
public boolean isValidBST(TreeNode node) {
        if(node == null) return true;
        TreeNode upper = null, lower = null;
        Queue<TreeNode[]>q = new LinkedList<>();
        q.add(new TreeNode[]{lower, node, upper});
        while(q.size()!=0){
            TreeNode[] t = q.poll();
            TreeNode l = t[0];
            TreeNode n = t[1];
            TreeNode u = t[2];
            if(n == null) continue;
            if(u!= null && n.val >= u.val) return false;
            if(l!= null && n.val <= l.val) return false;
            q.add(new TreeNode[]{l, n.left, n});
            q.add(new TreeNode[]{n, n.right, u});
        }
        return true;
    }   

File: PastYears/Leetcode75/98. Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool flag = false;
    bool check(TreeNode* root, TreeNode* lower, TreeNode* upper){
        if(root == NULL) return true;
        
        if(upper != NULL and root->val >= upper->val) return false;
        if(lower != NULL and root->val <= lower->val) return false;
        
        bool left = check(root->left, lower, root);
        bool right = check(root->right, root, upper);
        return left&&right;
    }
    
    bool isValidBST(TreeNode* root) {
        TreeNode *upper = NULL;// <=> +infinity
        TreeNode *lower = NULL;// <=> -infinity
        return check(root, lower, upper);
    }
};

/**/
class Solution {
public:
    bool flag = false;
    bool check(TreeNode* root, TreeNode* lower, TreeNode* upper){
        if(root == NULL) return true;
        
        if( (upper!=NULL && lower != NULL && root->val < upper->val && root->val > lower->val) //for the node in the middle of the tree
           or (upper == NULL and lower == NULL) // for the root node
           or (lower!=NULL and root->val > lower->val and upper == NULL) // for the right most node
           or (upper != NULL and root->val < upper->val and lower == NULL) // for the left most node
         ) {
            bool left = check(root->left, lower, root);
            bool right = check(root->right, root, upper);
            return left&&right;
        }else 
            return false; 
    }
    
    bool isValidBST(TreeNode* root) {
        TreeNode *upper = NULL;// <=> +infinity
        TreeNode *lower = NULL;// <=> -infinity
        return check(root, lower, upper);
    }
};
